code/datums/progressbar.dm

```diff
@@ -13,34 +16,60 @@
 	if (goal_number)
 		goal = goal_number
 	bar = image('icons/effects/progessbar.dmi', target, "prog_bar_0")
+	bar.alpha = 0
 	bar.appearance_flags = APPEARANCE_UI_IGNORE_ALPHA
 	bar.plane = HUD_PLANE
 	bar.layer = HUD_ABOVE_ITEM_LAYER
-	bar.pixel_y = WORLD_ICON_SIZE
 	src.user = user
 	if(user)
 		client = user.client
 
-/datum/progressbar/Destroy()
-	if (client)
-		client.images -= bar
-	qdel(bar)
-	. = ..()
+	LAZYINITLIST(user.progressbars)
+	LAZYINITLIST(user.progressbars[bar.loc])
+	var/list/bars = user.progressbars[bar.loc]
+	bars.Add(src)
+	listindex = bars.len
+	animate(bar, pixel_y = 32 + (PROGRESSBAR_HEIGHT * (listindex - 1)), alpha = 255, time = 5, easing = SINE_EASING)
 
 /datum/progressbar/proc/update(progress)
-//	log_debug("Update [progress] - [goal] - [(progress / goal)] - [((progress / goal) * 100)] - [round(((progress / goal) * 100), 5)]")
-
 	if (!user || !user.client)
 		shown = 0
 		return
 	if (user.client != client)
 		if (client)
 			client.images -= bar
-			shown = 0
-		client = user.client
+		if(user.client)
+			user.client.images += bar
 
 	progress = Clamp(progress, 0, goal)
 	bar.icon_state = "prog_bar_[round(((progress / goal) * 100), 5)]"
 	if (!shown && user.get_preference_value(/datum/client_preference/show_progress_bar) == GLOB.PREF_SHOW)
 		user.client.images += bar
 		shown = 1
+
+/datum/progressbar/proc/shiftDown()
+	--listindex
+	var/shiftheight = bar.pixel_y - PROGRESSBAR_HEIGHT
+	animate(bar, pixel_y = shiftheight, time = 5, easing = SINE_EASING)
+
+/datum/progressbar/Destroy()
+	for(var/I in user.progressbars[bar.loc])
+		var/datum/progressbar/P = I
+		if(P != src && P.listindex > listindex)
+			P.shiftDown()
+
+	var/list/bars = user.progressbars[bar.loc]
+	bars.Remove(src)
+	if(!bars.len)
+		LAZYREMOVE(user.progressbars, bar.loc)
+	animate(bar, alpha = 0, time = 5)
+	addtimer(CALLBACK(src, .proc/qdel_bar, bar), 5)
+	. = ..()
+
+/datum/progressbar/proc/qdel_bar(bar)
+	if(client)
+		client.images -= bar
+	qdel(bar)
```

Comment:
You need to null out the `bar` var too otherwise you'll cause GC failures and lag.
```suggestion
	qdel_null(bar)
```
(You need to do the above suggestion of "don't pass it as a param" for this snippet to work)